staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -300,7 +305,24 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
 }
 
 // Assumes that lock is already held for write.
+// Use dynamic size cache. If all cached events occur inside eventFreshDuration, increase cache capacity by 2x.
+// Decrease cache capacity by 2x when recent half events occur outside eventFreshDuration.
 func (w *watchCache) updateCache(event *watchCacheEvent) {
+	if w.endIndex == w.startIndex+w.capacity && event.RecordTime.Sub(w.cache[w.startIndex%w.capacity].RecordTime) < eventFreshDuration {
+		w.makeNewCache(w.capacity * 2)
+		w.capacity *= 2
+	} else if w.endIndex-w.startIndex < w.capacity/2 {
+		if w.capacity/2 > 0 {
+			w.makeNewCache(w.capacity / 2)
+			w.capacity /= 2
+		}
+	} else if event.RecordTime.Sub(w.cache[(w.endIndex-w.capacity/2)%w.capacity].RecordTime) > eventFreshDuration {
+		if w.capacity/2 > 0 {
+			w.startIndex = w.endIndex - w.capacity/2
+			w.makeNewCache(w.capacity / 2)
+			w.capacity /= 2
+		}
+	}
```

Comment:
Please encapsulate this resizing logic into a separate function - it would make testing simpler.